### PR TITLE
Fix PHP 8.0 issues

### PR DESCRIPTION
**Pull request by frostmakk**
Fix for issue #109

### CODE REVIEW

**Comment by RussW** [comment]
I can't test on PHP8.0 at the moment, but no adverse errors observed on PHP7.4.12 and mysqli connect errors report as expected still (funny it was only in mysqli, mysql didn't have the connection param)

I'm happy for this to progress as is. Thanks @frostmakk for your quick response.

**Comment by toivo** [comment]
The FPA script runs now in PHP 8.0.0 but reports that PHP does not support Joomla 3.9.23 and that the Confidence level is F.
**3.9.23:**
3 9 23-support
3 9 23-confidence

**4.0.0:**
4 0 0-support
4 0 0-confidence

The test environment was Windows 10, Wampserver 3.2.4 with PHP 8.0.0.

After PHP is reverted to **7.4.13**, the support and confidence for 3.9.23 returned to normal:

3 9 23-PHP7 4 13-support
3 9 23-PHP7 4 13-confidence

**Comment by sozzled** [comment]
Confidence level F ... hmm ... actually that's "logical".

The technical requirements documentation does not say that J! 3.x is compliant with PHP 8.  Further, when I asked for information about the technical requirements for J! 4, I was told it was none-of-my-business and that the requirements document would be "updated" closer to the actual release of J! 4 stable (see https://forum.joomla.org/viewtopic.php?f=7&t=982446).  So, basically, the FPA report is accurate in the sense that it matches what it finds against the technical requirements as they are published.

I don't think one can have it both ways:.

Nice to see that the FPA runs in a PHP 8 environment.  Thanks for testing.

**Comment by RussW** [comment]
@toivo could I ask a favour please, could you edit your FPA at line 3669 (maxPHP) to "8.0.0" (from 7.4.9 I think it was) and see if that fixes your current observation?

As @sozzled has eluded to, at the time PHP8 was not production and was not listed in the requirements, assuming that J3.9.xx is truly compatible with PHP8 we should be good to go, but there are quite some differences from v7 and a few backward compat changes.

**Comment by sozzled** [comment]
I have no idea (and I don't know if anyone knows the answer) if the J! framework or CMS is compatible with PHP 8.  I don't even know what version of the J! framework sits behind J!4 (I asked the question a couple of days ago but there's been no response).  Certainly, I'm aware that some libraries (included with J! 3.x and used by third-party developers) are incompatible with some versions of PHP 7.4 and there's no intention by the J! developers to address those matters that lie "upstream" from the project.  Only time will tell, I guess.

Not sure if that helps much.  I've been struggling to get some sense of where the J! framework sits at the moment and that could be a problem in working on the "confidence level".  From what I've looked at with the PHP code that runs the FPA, thinss look pretty good to me.  :sunglasses:

**Comment by RussW** [comment]
It's my understanding that the J! Framework was de-coupled from the CMS at J3.2 and is now independently developed & maintained but namespaces, functions and methods are kept in sync and the framework is just another "library" per sai' now, that can be used by the CMS, Extensions and API's alike, making it much more flexible.

Again, if memory serves me well, this also means that at J!4 there is now also no platform, for this reason. The framework has taken over those dependancies and functions.

Current J!4 minimum PHP is v7.2/3 due to several bugs in V7.0 & v7.1 that would generate huge backward compat issues when moving moving to v8, namely some empty objects that were assigned, no longer need to be and the use of curly-braces is being retired for many methods to avoid confusion and parsing by reference (which was started to be removed on PHP7), on  a good note many STRICT errors have been reclassified and will no-longer cause fatals.  

We have a few arrays that are "parsed by reference" that may need attention if PHP8 becomes stricter than it's current aspirations, but most good/pro extensions and the CMS stopped using parsing by reference pre-PHP7.

Some error routines may need to be re-written in J! CMS to adequately work with the new error_reporting mechanisms, but shouldn't fatal error as PHP8 is just re-assigning their level on the fly.

These issues with FPA are purely procedural issues, not PHP incompatibility issues as such, not saying we wont bump in t some in later releases of PHP, but at this time, FPA is effectively PHP8 compatible, just not handling v8 well.

Hope that helps somewhat and sheds some light rather than fog.

**Comment by toivo** [comment]
@RussW after the maxPHP entry is changed from 7.5.0 to 8.0.0 the FPA reports that PHP support of Joomla 3.9.3 is back to 'Yes' and the confidence level goes to B, the same values as Joomla 4 Beta6-dev in my test environment using Wampserver 3.2.4 on Windows 10.

**Comment by RussW** [comment]
Thank you @toivo, really appreciate your time, patience and assistance with this issue. I've now "faked" running PHP8 on my testbed server and all checks out fine as well.

@frostmakk : Should we now simply assume that J3.9 is compatible with PHP8 and change the known '**_maxPHP_**' value to "**8.0.0**"? I'm leaning toward doing this until proven wrong. 

If you agree, would you be happy to adjust your PR and also include changes to the maxPHP version for PHP at J3.9 to "8.0.0" at line 3669. I guess we need to prepare for later v8 releases also.

**Comment by frostmakk** [comment]
I'll do some more tests with 3.9 on 8.0 later today to check that all versions of 3.9 behaves.
I'm ok with assuming compatibility, but would prefer to set the max value to something like 8.0.25 so we don't have to do it all over again "tomorrow" when the first 8.0 bugfix is released. 
If that's ok I'll update the PR this evening. (local time 😄 )

**Comment by RussW** [comment]
Sounds like a plan to me, maybe 8.1.0 just to cover the first minor release after patches?

**Comment by sozzled** [comment]
Thanks for your information about "framework" vs "platform".  I was aware that the platform (at one point) went off in its own direction but that its now been re-absorbed back into the framework project.  It still leaves me a little confused w.r.t. the roadmap (and, therefore, whether J! 4 uses the same framework as J! 3.x) but that's a side issue.  As long as the FPA can use the J! API calls then it's neither here nor there ... unless Framework v 2.0 changes things, that is.

It still would be nice to get some information about where the J! project is heading and to address the unanswered question I asked on the J! forum (https://forum.joomla.org/viewtopic.php?f=803&t=983281).

But, to return to how the FPA may help people who have questions/concerns about PHP 8.0 (and the inference—from an "F" confidence score—that J! x.y is incompatible with PHP 8), I can only refer to what we know:  the technical requirements.  Until the technical requirements are updated (and that's unlikely to occur in the near future), based on the developers' assurances that J! x.y is compatible with PHP 8, everything we say or do here is a little "best guess", I think.  I'm not making any criticism; I'm just putting it "out there".

In summary, it's nice to get the **_FPA_** compatible with PHP 8.  Discussing J! CMS/Framework compatibility with PHP is outside our scope, I think.

**Comment by RussW** [comment]
Awesome work @frostmakk, thank you for your time and testing - looks all good to me

**Comment by frostmakk** [comment]
Environment checks added to PR.
The very basic tests I did was: If Joomla installs, and i I can open the administrator CP without Joomla crashing, it is compatible.
Any version before 3.9.13 could not be installed on neither PHP 8.0 nor PHP 7.4

**Comment by frostmakk** [comment]
I also added the STFU directive to the rest of the Date() functions to avoid errors like this pushing the interesting errors out of sight:
_Warning: date() [function.date]: It is not safe to rely on the system's timezone settings. You are *required* to use the date.timezone setting or the date_default_timezone_set() function._

**Comment by frostmakk** [comment]
Discovered another feature in PHP 8.0 during testing. It's not so forgiving as earlier versions when it comes to connection problems to database, and throws a Fatal error at first glimpse of trouble.
Had to add an extra error trap to the PR.

**Comment by frostmakk** [comment]
It can be hard to see the difference between 1 (one) and l (small letter L) in the randomly generated table prefixes.
(Once I wasted an entire day trying to figure out what was wrong)
This last push to the PR is just a split of the table count in the screen form to aid people without eagle eyes.



image

**Comment by frostmakk** [comment]
Release info added.
I believe it is safe to proceede with this now.
@RussW   If You could do a final verification and merge, then I'll release 1.6.2 if everybody is happy.

**Comment by sozzled** [comment]
It would have been nice if the code changes had been commented with name/date information but, as far as I can tell (without having access to a PHP 8 development lab), the code looks syntactically correct and shouldn't cause compiler issues.  I'll give it a run-through under PHP 7 conditions later; irrespective of what I find between now and then, please release the v1.6.2 (we can always revert things later if it blows up).  Cheers :smiley:

**Comment by RussW** [comment]
Thanks @frostmakk, I'll be building PHP8 on our Test/Dev Servers over the weekend, but at the moment all seems fine on PHP7.4.13 on Apache and nginx.
